R/conjunct_stopifnot_linter.R

```diff
@@ -0,0 +1,39 @@
+#' Force && conditions in stopifnot() to be written separately
+#'
+#' `stopifnot()` accepts any number of tests, so sequences like
+#'   `stopifnot(x && y)` are better written as `stopifnot(x, y)` because
+#'   the error message in the latter case is more useful.
+#'
+#' @evalRd rd_tags("stopifnot_and_condition_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+conjunct_stopifnot_linter <- function() {
+  Linter(function(source_file) {
+    # need the full file to also catch usages at the top level
+    if (length(source_file$full_xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$full_xml_parsed_content
+
+    xpath <- "
```

Comment:
I also need to add a TODO for R>=4.0 not to lint in this case:

```
stopifnot(`Descriptive reason` = A && B)
```

possibly optionally

Reply:
Hmm. Could still be written as
```
stopifnot(
  `Descriptive reason, A failed` = A,
  `Descriptive reason, B failed` = B
)
```

Reply:
Right, but it can be tedious... consider

```
stopifnot(
  `x is a logical scalar` = length(x) == 1 && is.logical(x) && !is.na(x),
  `y is a logical scalar` = length(y) == 1 && is.logical(y) && !is.na(y), 
)
```

So I think it should be optional.

note that this is equivalent to allowing

```
assertthat::assert_that(length(x) == 1, is.logical(x), !is.na(x), msg="x is a logical scalar")
```

instead of requiring several `assert_that()` calls with a unique `msg`